src/katsdpmodels/primary_beam.py

```diff
@@ -0,0 +1,256 @@
+        self.unit = unit
+        self.required = required
+
+
+class PrimaryBeam(models.SimpleHDF5Model):
+    r"""Base class for primary beams.
+
+    The phase of the electric field at a point increases with time i.e., the
+    phasor is
+
+    .. math:: e^{(\omega t - kz)i}
+
+    The "ideal" complex voltage is multiplied by the beam model to obtain the
+    measured voltage.
+
+    If you stick your right arm out and left hand up, then pretend to be an
+    antenna (lean back and look at the sky) then they correspond to the
+    directions of positive horizontal and vertical polarization respectively
+    (the absolute signs don't matter but the signs relative to each other do
+    for cross-hand terms).
+
+    An instance of this class describes either a single antenna or an average
+    of multiple antennas; to do antenna-specific modelling you will need a
+    separate instance of this class per antenna.
+    """
+
+    model_type: ClassVar[Literal['primary_beam']] = 'primary_beam'
+
+    def resolution(self, frequency: u.Quantity) -> float:
+        """Approximate spatial resolution of the model, in units of projected coordinates.
+
+        Sampling a grid at significantly higher resolution than this will have
+        diminishing returns compared to sampling at this resolution and
+        interpolating.
+        """
+        raise NotImplementedError()
+
+    @property
+    def frequency_range(self) -> Tuple[u.Quantity, u.Quantity]:
+        """Minimum and maximum frequency covered by the model."""
+        raise NotImplementedError()
+
+    @property
+    def frequency_resolution(self) -> u.Quantity:
+        """Approximate frequency resolution of the model.
+
+        Sampling at significantly higher spectral resolution than this will
+        have diminishing returns compared to sampling at this resolution and
+        interpolating.
+        """
+        raise NotImplementedError()
+
+    def min_radius(self, frequency: u.Quantity) -> float:
+        """Maximum distance from the pointing centre at which model has full coverage."""
+        raise NotImplementedError()
+
+    def max_radius(self, frequency: u.Quantity) -> float:
+        """Maximum distance from the pointing centre at which model has any coverage."""
+        raise NotImplementedError()
+
+    @property
+    def is_circular(self) -> bool:
+        """Whether this model is circularly-symmetric about the pointing centre."""
+        raise NotImplementedError()
+
+    @property
+    def is_unpolarized(self) -> bool:
+        """Whether this model ignores polarization.
+
+        If true, it is guaranteed that the Jones matrices describing the beam are
+        scaled identity matrices.
+        """
+        raise NotImplementedError()
+
+    @property
+    def parameters(self) -> Sequence[Parameter]:
+        """Parameters of the model."""
+        raise NotImplementedError()
+
+    def sample(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,   # noqa: E741
```

Comment:
In my head I had the output axis order as frequency, m, l (so that the horizontal coordinate is the fastest-varying, as is conventional for images), although I see the documentation doesn't actually say that.

Reply:
Ah right, then I would leave the input parameters as is, because `(m, l)` is going to throw some folks on the input side, and then the input is at least the reverse of the output.